Source/ARTConnection.m

```diff
@@ -303,3 +291,49 @@ + (instancetype)newWithConnectionEvent:(ARTRealtimeConnectionEvent)value {
 }
 
 @end
+
+@implementation ARTConnectionRecoveryKey
+
+- ( NSString *)jsonString{
+    NSError *error;
+    NSDictionary *object = @{
+        @"msgSerial": @(_msgSerial),
+        @"connectionKey": _connectionKey,
+        @"serials": _channelSerials ? _channelSerials : @{}
+    };
+    
+    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:object
+                                                       options:0
+                                                         error:&error];
+    if (error) {
+        return nil;
+    } else {
+        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+    }
+    return nil;
+}
+
++(ARTConnectionRecoveryKey *)fromJsonString:(NSString *)json{
+        NSData* jsonData = [json dataUsingEncoding:NSUTF8StringEncoding];
+
+        NSError *error = nil;
+        NSDictionary  *object = [NSJSONSerialization
+                                 JSONObjectWithData:jsonData
+                                 options:0
+                                 error:&error];
+
+        if(!error) {
+            ARTConnectionRecoveryKey *recoveryKey = [[ARTConnectionRecoveryKey alloc] init];
+            recoveryKey.msgSerial = [[object valueForKey:@"msgSerial"] longLongValue];
+            recoveryKey.connectionKey = [object valueForKey:@"connectionKey"];
+            recoveryKey.channelSerials = [object valueForKey:@"serials"];
+            return recoveryKey;
+            
+        } else {
+            NSLog(@"Error when parsing JSON to create recovery key : %@", error);
```

Comment:
I'm a little bit concerned that this could become a supportability concern in the future. My preference would be for this to fail early to assist customers debugging their apps. Akin and aligned with my comment on your `jsonString` implementation, I think you should raise an exception here. Alongside that, I think this parsing step should happen when the client options containing this key is presented to the client instance constructor.

Reply:
The throwing exception for this is done as part of 5d9efe9547889708f30c394c420c6e14d43bd7ba

> Alongside that, I think this parsing step should happen when the client options containing this key is presented to the client instance constructor.

That should already be happening. If you check 

`- (instancetype)initWithOptions:(ARTClientOptions *)options 
`
in ARTRealtime class
```
 if (options.recover) { //RTN16f
            ARTConnectionRecoveryKey *const recoveryKey = [ARTConnectionRecoveryKey fromJsonString:options.recover];
            self.msgSerial = recoveryKey.msgSerial;
            self.connection.latestMessageSerial = self.msgSerial;
            [self recoverChannels: recoveryKey]; //RTN16j
        } else {
            self.msgSerial = 0;
            self.connection.latestMessageSerial = 0;
        }
```